zanata-war/src/main/java/org/zanata/transaction/TransactionUtil.java

```diff
@@ -0,0 +1,162 @@
+
+            transactionActive = transaction.isActiveOrMarkedRollback()
+                    || transaction.isRolledBack(); // TODO: temp workaround,
+                                                   // what should we really do
+                                                   // in this case??
+            newTransactionRequired =
+                    isNewTransactionRequired(transactionActive);
+            userTransaction = newTransactionRequired ? transaction : null;
+        } catch (IllegalStateException e) {
+            // for shutdown case, when we can't get the tx object because the
+            // event context is gone
+            // but we should still check if a tx is required and fail
+            // accordingly if it is
+            newTransactionRequired = isNewTransactionRequired(false);
+            if (newTransactionRequired) {
+                throw e;
+            }
+        }
+
+        try {
+            if (newTransactionRequired) {
+                log.debug("beginning transaction");
+                userTransaction.begin();
+            }
+
+            R result = function.call();
+            if (newTransactionRequired) {
+                if (transaction.isMarkedRollback()) {
+                    log.debug("rolling back transaction");
+                    userTransaction.rollback();
+                } else {
+                    log.debug("committing transaction");
+                    userTransaction.commit();
+                }
+            }
+            return result;
+        } catch (Exception e) {
+            if (newTransactionRequired
+                    && userTransaction.getStatus() != Status.STATUS_NO_TRANSACTION) {
+                if (isRollbackRequired(e, true)) {
+                    log.debug("rolling back transaction");
+                    userTransaction.rollback();
+                } else {
+                    log.debug("committing transaction after ApplicationException(rollback=false):"
+                            + e.getMessage());
+                    userTransaction.commit();
+                }
+            }
+            throw e;
+        }
+    }
+
+    /**
+     * Same as {@link TransactionUtil#runInTransaction(Callable)} but for
+     * Runnables (functions that don't return anything)
+     *
+     * @param function
+     *            The function (in the form of a Runnable) to execute.
+     * @throws Exception
+     *             Exception (if any) thrown by the given function.
+     * @see TransactionUtil#runInTransaction(Callable)
+     */
+    public static void runInTransaction(Runnable function) throws Exception {
+        runInTransaction(() -> {
+            function.run();
+            return null;
+        });
+    }
+
+    protected static boolean
+            isNewTransactionRequired(boolean transactionActive) {
+        return !transactionActive;
+    }
+
+    public static boolean isRollbackRequired(Exception e, boolean isJavaBean) {
```

Comment:
Looks like we only call this method with isJavaBean as true. May clean this up to only accept e as parameter